src/server.rs

```diff
@@ -58,65 +54,567 @@ use protocol::{
-    storage: Arc<Box<Storage>>,
+impl<C: CommandCreatorSync> SccacheServer<C> {
+    pub fn new(port: u16) -> io::Result<SccacheServer<C>> {
+        let core = Core::new()?;
+        let handle = core.handle();
+        let addr = SocketAddrV4::new(Ipv4Addr::new(127, 0, 0, 1), port);
+        let listener = TcpListener::bind(&SocketAddr::V4(addr), &handle)?;
+
+        // Prepare the service which we'll use to service all incoming TCP
+        // connections.
+        let (tx, rx) = mpsc::channel(100);
+        let pool = CpuPool::new(20);
+        let storage = storage_from_environment(&pool);
+        let (wait, info) = WaitUntilZero::new();
+        let service = SccacheService::new(storage, core.handle(), pool, tx, info);
 
-    /// Server statistics.
-    stats: ServerStats,
+        Ok(SccacheServer {
+            core: core,
+            listener: listener,
+            rx: rx,
+            service: service,
+            timeout: Duration::from_millis(DEFAULT_IDLE_TIMEOUT),
+            wait: wait,
+        })
+    }
 
-    /// True if the server is actively shutting down.
-    shutting_down: bool,
+    /// Configures how long this server will be idle before shutting down.
+    #[allow(dead_code)]
+    pub fn set_idle_timeout(&mut self, timeout: Duration) {
+        self.timeout = timeout;
+    }
+
+    /// Set the `force_recache` setting.
+    #[allow(dead_code)]
+    pub fn set_force_recache(&mut self, force_recache: bool) {
+        self.service.force_recache = force_recache;
+    }
+
+    /// Set the `force_recache` setting.
+    #[allow(dead_code)]
+    pub fn set_storage(&mut self, storage: Arc<Storage>) {
+        self.service.storage = storage;
+    }
+
+    /// Returns a reference to the thread pool this server will run work on
+    #[allow(dead_code)]
+    pub fn pool(&self) -> &CpuPool {
+        &self.service.pool
+    }
+
+    /// Returns a reference to the command creator this server will use
+    #[allow(dead_code)]
+    pub fn command_creator(&self) -> &C {
+        &self.service.creator
+    }
+
+    /// Returns the port that this server is bound to
+    #[allow(dead_code)]
+    pub fn port(&self) -> u16 {
+        self.listener.local_addr().unwrap().port()
+    }
+
+    /// Returns a handle to the event loop that this server is bound to.
+    #[allow(dead_code)]
+    pub fn handle(&self) -> Handle {
+        self.core.handle()
+    }
+
+    /// Runs this server to completion.
+    ///
+    /// If the `shutdown` future resolves then the server will be shut down,
+    /// otherwise the server may naturally shut down if it becomes idle for too
+    /// long anyway.
+    pub fn run<F>(self, shutdown: F) -> io::Result<()>
+        where F: Future,
+    {
+        self._run(Box::new(shutdown.then(|_| Ok(()))))
+    }
+
+    fn _run<'a>(self, shutdown: Box<Future<Item = (), Error = ()> + 'a>)
+                -> io::Result<()>
+    {
+        let SccacheServer { mut core, listener, rx, service, timeout, wait } = self;
```

Comment:
Yeah I figured it'd provide a good entry point to configure tests and other such style-shims but I'd also be totally fine just inlining above in `start-server`